core-rust/state-manager/src/store/node_ancestry_resolver.rs

```diff
@@ -0,0 +1,486 @@
+pub struct NodeAncestryResolver {}
+
+impl NodeAncestryResolver {
+    /// Resolves the [`SubstateNodeAncestryRecord`] for all newly-created Nodes found in the given
+    /// [`SubstateChange`]s.
+    /// The resolution considers the trees (or rather: potentially-incomplete bottom-up tree
+    /// fragments) constructed within the given [`SubstateChange`]s, and the top fragments of these
+    /// trees already-existing in the given [`SubstateNodeAncestryStore`].
+    /// API note: the results are grouped by [`SubstateNodeAncestryRecord`] (since it may be shared
+    /// by many child [`NodeId`]s).
+    pub fn batch_resolve<S: SubstateNodeAncestryStore>(
+        ancestry_store: &S,
+        substate_changes: &[SubstateChange],
+    ) -> impl Iterator<Item = (Vec<NodeId>, SubstateNodeAncestryRecord)> {
+        // Gather the Nodes owned by upserted parent Substates (using `IndexedScryptoValue`).
+        // This effectively builds a forest of upserted nodes, using a "parent to child-list map" representation.
+        let directly_owned_node_sets =
+            Self::extract_owned_node_sets(substate_changes).collect::<Vec<_>>();
+
+        // Prune the child-lists in the above forest so that they only contain nodes which are parents of some other nodes.
+        // This limits further computation, since we only need to know a root of each parent anyway.
+        let directly_owned_parent_sets =
+            Self::clone_parents_from_node_sets(&directly_owned_node_sets);
+
+        // Split the above "parents-only" forest into separate trees and flatten them.
+        // This way we obtain all parents owned by each topmost parent within `SubstateChange`s.
+        let transitively_owned_parent_sets =
+            Self::propagate_transitively_owned_sets(directly_owned_parent_sets).collect::<Vec<_>>();
+
+        // Resolve the existing roots of the topmost parents found above.
+        // This queries the DB and then applies the "any Node without DB entry is a root" rule.
+        let topmost_parent_roots = Self::batch_get_existing_root_or_same_substate(
+            ancestry_store,
+            transitively_owned_parent_sets
+                .iter()
+                .map(|(parent, _)| parent),
+        );
+
+        // Combine the `transitively_owned_parent_sets` (lower parts of the trees, constrcuted from
+        // `SubstateChange`s) and `topmost_parent_roots` (roots of the trees, loaded from DB).
+        let parent_to_root = transitively_owned_parent_sets
+            .iter()
+            .zip(topmost_parent_roots)
+            .flat_map(|((topmost_parent, transitively_owned_parents), root)| {
+                let root_for_topmost = root.clone();
+                transitively_owned_parents
+                    .iter()
+                    .map(move |owned_parent| (*owned_parent, root.clone()))
+                    // interesting gotcha: the topmost parent is also "a parent" (and won't be on any child-list)
```

Comment:
> What if the top-most parent is itself a root here?

Absolutely not a problem: the `parent_to_root` map will then simply contain an entry like `topmost_parent=7 -> root=7`. And the iteration at the end will query it out, and it will be a correct answer.

But I agree: if it is itself a root, we can totally skip it from `parent_to_root`, and the iteration at the end will correctly assume it is a root then.

So I applied this change and added extra comments to the gotcha.

I adjusted `newly_created_root_nodes_are_not_recorded_in_index` to show that, as you asked 👍 
(and it is passing for both versions, before and after the change)